docs/guides/relay.md

```diff
@@ -0,0 +1,394 @@
+---
+title: Relay
+---
+
+# Relay Guide
+
+## What is Relay?
+
+The relay spec defines some interfaces that GraphQL servers can follow to allow
+clients to interact with them in a more efficient way. The spec makes two
+core assumptions about a GraphQL server:
+
+1. It provides a mechanism for refetching an object
+2. It provides a description of how to page through connections.
+
+You can read more about the relay spec
+[here](https://relay.dev/docs/en/graphql-server-specification/).
+
+### Relay implementation example
+
+Suppose we have the following type:
+
+```python
+@strawberry.type
+class Fruit:
+    name: str
+    weight: str
+```
+
+We want it to have a globally unique ID, a way to retrieve a paginated results
+list of it and a way to refetch if if necessary. For that, we need to inherit it
+from the `Node` interface and implement its abstract methods: `resolve_id`,
+and `resolve_nodes`.
+
+```python
+from strawberry import relay
+
+
+@strawberry.type
+class Fruit(relay.Node):
+    code: relay.NodeID[str]
+    name: str
+    weight: float
+
+    @classmethod
+    def resolve_nodes(
+        cls,
+        *,
+        info: Info,
+        node_ids: Optional[Iterable[str]] = None,
+        required: bool = False,
+    ):
+        if node_ids is not None:
+            return [fruits[nid] if required else fruits.get(nid) for nid in node_ids]
+
+        return fruits.values()
+
+
+# Assume we have a dict mapping the fruits code to the Fruit object itself
+fruits: Dict[int, Fruit]
```

Comment:
I think the type annotation here should be `Dict[str, Fruit]` - because the node ID field in the example is str.

Reply:
`int` is actually fine because the globalid will do `str(id)` on it. Most ids are usually ints/uuids

Reply:
> So I think what I mean is either the dictionary should be str -> Fruit or the resolve_nodes class should have an int(nid) cast before using it as a key.

Indeed if the id was set to `str` it would work out of the box. But we can't just do `int(nid)` when passing the ids to `resolve_nodes` because we can't actually know the type of the ids when we resolve them.

e.g. You might have `FooType:1` and that `1` might be an integer or really a string containing the character `"1"`.

I don't know if there's a better way to workaround this, but I decided to go the easy way in this initial implementation and let the implementer cast the value. In this case, since you know that the id is an integer, you can do `int(nid)` inside the `resolve_nodes` function, like: `[fruits[int(nid)] if required else fruits.get(int(nid)) for nid in node_ids]`

But I'm open to suggestions

Reply:
I think the way you have it in the interface is fine. I don't think relay specifies any particular format for the ID so as you say you can't know.  Its fine to leave it up to the implementer.

I just mean in the example in the docs it would make it clearer to either have the `int(nid)` cast in the code, or make the dict keys `str` - as it is type checkers will error on the lookup. I've just noticed in the very similar code in the tests - `schema.py` line 122, the ids are cast to str in the lookup dictionary which makes sense.

Reply:
Oh, regarding the docs you are totally right, will fix it! Thanks for pointing it out

Reply:
Fixed!